odoo/addons/base/tests/test_base.py

```diff
@@ -52,13 +64,275 @@ def test_05_safe_eval_forbiddon(self):
+
+        code = """
+        w = WeirdClass()
+        w.addAttr("obj", print)
+        w.addAttr("obj_attr", ('__self__', ))
+        w.getattrib("__self__").exec("import this")
+        """
+
+        with self.assertRaises(ValueError):
+            safe_eval(cleandoc(code), globals_dict={
+                      "StrictStr": StrictStr, "WeirdClass": WeirdClass, "print": print}, mode="exec")
+
+    def test_10_format_should_be_denied(self):
+        # format is forbidden on strings
+        # because it can be used to leak data through dunders
+
+        codefmt = cleandoc(
+            """
+            fmt = "Hello {name}"
+            fmt.format(name="World")
+            """
+        )
+
+        codefmt2 = cleandoc(
+            """
+            fmt = "Hello {}"
+            fmt.format("World")
+            """
+        )
+
+        codefmt3 = cleandoc(
+            """
+            fmt = "Hello {name}"
+            fmt.format_map({"name": "World"})
+            """
+        )
+
+        codeCfmt = cleandoc(
+            """
+            fmt = "Hello %s"
+            fmt % "World"
+            """
+        )
+
+        codeNewFmt = cleandoc(
+            """
+            fmt = f"Hello {'World'}"
+            """
+        )
+
+        markup_format = "Markup('Hello {}'.format('World'))"
+
+        with self.assertRaises(ValueError):
+            safe_eval(codefmt, mode="exec")
+
+        with self.assertRaises(ValueError):
+            safe_eval(codefmt2, mode="exec")
+
+        with self.assertRaises(ValueError):
+            safe_eval(codefmt3, mode="exec")
+
+        with self.assertRaises(ValueError):
+            safe_eval(markup_format, mode="exec", globals_dict={"Markup": Markup})
+
+        # Those type of format are allowed
+        safe_eval(codeCfmt, mode="exec")
+        safe_eval(codeNewFmt, mode="exec")
+
+        # Check that format method on other objects is allowed
+        class SomeObject:
+            def format(*args, **kwargs):    # noqa: A003 # pylint: disable=no-method-argument
+                ...
+
+        code = cleandoc(
+            """
+            fmt = SomeObject()
+            fmt.format(name="World")
+            """
+        )
+
+        safe_eval(code, globals_dict={"SomeObject": SomeObject}, mode="exec", sandboxed_types=(SomeObject,))
+
+    def test_12_use_of_denied_ast_nodes(self):
+        # Case 1: use of ast.Call with safe_eval in "math" mode
+        with self.assertRaises(ValueError):
+            safe_eval("foo()", ast_subset=_MATH_NODES | {ast.Load})
+
+        # Case 2: Forcing ast.ImportFrom inside of the sandbox
+        with self.assertRaises(ValueError):
+            safe_eval("from random import randint", ast_subset=_ALLOWED_NODES | {ast.ImportFrom})
+
+
+    def test_13_wrap_module(self):
+        # Everything in a module is visible, but only the allowed objects are accessible
+
+        with self.assertRaises(ValueError):
+            safe_eval("datetime.sys")
```

Comment:
does it crash because you check it or because `datetime` is not in the evaluation context ?

because of that I'm now afraid that other test fail for the wrong reason :/ 

so we might need to check the message of all check

![image](https://github.com/odoo/odoo/assets/54316226/6168d543-fc45-4e67-8ad1-f1a6993e42d7)